lib.sh

```diff
@@ -829,105 +829,73 @@ rm -r "$GPGDIR"
 rm -f releases
 }
 
-# Initial download of script in ../static
-# call like: download_static_script name_of_script
-download_static_script() {
-    # Get ${1} script
-    rm -f "${SCRIPTS}/${1}.sh" "${SCRIPTS}/${1}.php" "${SCRIPTS}/${1}.py"
-    if ! { curl_to_dir "${STATIC}" "${1}.sh" "$SCRIPTS" || curl_to_dir "${STATIC}" "${1}.php" "$SCRIPTS" || curl_to_dir "${STATIC}" "${1}.py" "$SCRIPTS"; }
+# call like: download_script folder_variable name_of_script
+# e.g. download_script APP additional_apps
+# Use it for functions like download_static_script
+download_script() {
+    rm -f "${SCRIPTS}/${2}.sh" "${SCRIPTS}/${2}.php" "${SCRIPTS}/${2}.py"
+    if ! { curl_to_dir "${!1}" "${2}.sh" "$SCRIPTS" || curl_to_dir "${!1}" "${2}.php" "$SCRIPTS" || curl_to_dir "${!1}" "${2}.py" "$SCRIPTS"; }
     then
-        print_text_in_color "$IRed" "{$1} failed to download. Please run: 'sudo curl -sLO ${STATIC}/${1}.sh|.php|.py' again."
+        print_text_in_color "$IRed" "{$2} failed to download. Please run: 'sudo curl -sLO ${!1}/${2}.sh|.php|.py' again."
         print_text_in_color "$ICyan" "If you get this error when running the nextcloud-startup-script then just re-run it with:"
         print_text_in_color "$ICyan" "'sudo bash $SCRIPTS/nextcloud-startup-script.sh' and all the scripts will be downloaded again"
         exit 1
     fi
 }
 
-# Initial download of script in ../lets-encrypt
-# call like: download_le_script name_of_script
-download_le_script() {
-    # Get ${1} script
-    rm -f "${SCRIPTS}/${1}.sh" "${SCRIPTS}/${1}.php" "${SCRIPTS}/${1}.py"
-    if ! { curl_to_dir "${LETS_ENC}" "${1}.sh" "$SCRIPTS" || curl_to_dir "${LETS_ENC}" "${1}.php" "$SCRIPTS" || curl_to_dir "${LETS_ENC}" "${1}.py" "$SCRIPTS"; }
+# call like: run_script folder_variable name_of_script
+# e.g. run_script APP additional_apps
+# Use it for functions like run_static_script
+run_script() {
+    rm -f "${SCRIPTS}/${2}.sh" "${SCRIPTS}/${2}.php" "${SCRIPTS}/${2}.py"
+    if curl_to_dir "${!1}" "${2}.sh" "$SCRIPTS"
```

Comment:
Why do you use `!` here: `curl_to_dir "${!1}" "${2}.sh" "$SCRIPTS"`

Reply:
I think I get it, you need to do that because we can't use `$1` here due to the varaibles needed like APP, STATIC, LETS_ENC and so on.

Reply:
yes, that's right, couldn't explain it better 😅

Reply:
a different way would be printing the variabes out directly inside the run_static_script functions but I like it more this way

Reply:
OK, great. But from where did you get that idea? I would like the background of it. Is it consider stable to do like that, or could it be mixed up with some other stuff?

Actually, I would prefer using the real variable. Then  we could check with `if [ -n $APP ]; then...` to make it more failproof.

Reply:
it is called "indirect expansion"
here you can find an explanation:
http://ahmed.amayem.com/bash-indirect-expansion-exploration/#ipt_kb_toc_1527_3

Reply:
the other solution which is possible is this:
```
download_script() {
    rm -f "${SCRIPTS}/${2}.sh" "${SCRIPTS}/${2}.php" "${SCRIPTS}/${2}.py"
    if ! { curl_to_dir "${1}" "${2}.sh" "$SCRIPTS" || curl_to_dir "${1}" "${2}.php" "$SCRIPTS" || curl_to_dir "${1}" "${2}.py" "$SCRIPTS"; }
    then
        print_text_in_color "$IRed" "{$2} failed to download. Please run: 'sudo curl -sLO ${1}/${2}.sh|.php|.py' again."
        print_text_in_color "$ICyan" "If you get this error when running the nextcloud-startup-script then just re-run it with:"
        print_text_in_color "$ICyan" "'sudo bash $SCRIPTS/nextcloud-startup-script.sh' and all the scripts will be downloaded again"
        exit 1
    fi
}

download_static_script() {
download_script "$STATIC" "${1}"
}
```
vs
```
download_script() {
    rm -f "${SCRIPTS}/${2}.sh" "${SCRIPTS}/${2}.php" "${SCRIPTS}/${2}.py"
    if ! { curl_to_dir "${!1}" "${2}.sh" "$SCRIPTS" || curl_to_dir "${!1}" "${2}.php" "$SCRIPTS" || curl_to_dir "${!1}" "${2}.py" "$SCRIPTS"; }
    then
        print_text_in_color "$IRed" "{$2} failed to download. Please run: 'sudo curl -sLO ${!1}/${2}.sh|.php|.py' again."
        print_text_in_color "$ICyan" "If you get this error when running the nextcloud-startup-script then just re-run it with:"
        print_text_in_color "$ICyan" "'sudo bash $SCRIPTS/nextcloud-startup-script.sh' and all the scripts will be downloaded again"
        exit 1
    fi
}
download_static_script() {
download_script STATIC "${1}"
}
```
Which one do you like more?

Reply:
this is another explanation:
https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html